### PR TITLE
New: Add support for root: true in config files (fixes #2736)

### LINKED ISSUE #2736: Alter config lookup?
this was brought up in the chat room, and I want to make sure we don't lose the discussion.there still seems to be a lot of confusion around configuration look up and how far it does or should go up to the directory structure. We had up until this point I looked all the way up to the home directory for configuration files. We would only use a configuration file in the home directory if there were no other configuration files found in subdirectories.

the current way this works was implemented here: https://github.com/eslint/eslint/issues/963

there was a discussion about whether or not we should keep looking up the directory structure or if we should stop at the current working directory. We eventually resolved to go all the way up to the home directory: https://github.com/eslint/eslint/issues/939#issuecomment-44234512

The question now is whether or not that choice was correct, and if it was correct, do we want a way to allow people to opt-out and specify where they want the lookup to stop.

The suggestion in the chatroom was to add "root": true to a config to say " don't look up any further from here. "

What do people think?

### PR DESCRIPTION
**Pull request by IanVS**

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -229,6 +230,16 @@ function getLocalConfig(thisConfig, directory) {
             continue;
         }
 
+        // Check for root flag
+        if (localConfig.root === true) {
+            rootPath = path.dirname(localConfigFile);
+        }
+
+        // If root flag is set, don't consider file if it is above root
+        if (rootPath && localConfigFile.indexOf(rootPath) === -1) {
```
Comment (gyandeeps) [inline lib/config.js]: Please use https://github.com/domenic/path-is-inside module to figure out if a particular path is inside another path.
Certain times your condition would fail. This module talks about those conditions.

**Comment by gyandeeps** [comment]
Please update config file documentation for root option.

File: lib/config.js (hunk 1 above)
Comment (IanVS) [inline lib/config.js]: Thanks for the tip, that's good to know.  The main issue with indexOf seems to be that on Windows you can have the same path with different capitalizations.  However, in this case where the paths being compared are all returned from the same process, is that really a concern?  I'd hate to introduce a dependency if one isn't needed.

**Comment by IanVS** [comment]
Please feel free to suggest better wording for the documentation.

File: docs/user-guide/configuring.md [hunk 2]
```diff
@@ -372,6 +372,21 @@ your-project
   └── test.js
 ```
 
+By default, ESLint will look for configuration files in all parent folders up to the home directory. This can be useful if you want all of your projects to follow a certain convention, but can sometimes lead to unexpected results. To limit ESLint to a specific project, place `"root": true` inside the `eslintConfig` field of the `package.json` file or in the `.eslintrc` file at your project's root level.
+
+```js
```
Comment (gyandeeps) [inline docs/user-guide/configuring.md]: See if you can share an example on how the lookup would work.

**Comment by IanVS** [comment]
@gyandeeps How's this?

**Comment by gyandeeps** [comment]
Thanks @IanVS . Docs looks good.

File: docs/user-guide/configuring.md [hunk 3]
```diff
@@ -372,6 +372,36 @@ your-project
   └── test.js
 ```
 
+By default, ESLint will look for configuration files in all parent folders up to the home directory. This can be useful if you want all of your projects to follow a certain convention, but can sometimes lead to unexpected results. To limit ESLint to a specific project, place `"root": true` inside the `eslintConfig` field of the `package.json` file or in the `.eslintrc` file at your project's root level.
```
Comment (btmills) [inline docs/user-guide/configuring.md]: Nit: ESLint searches all the way up to the root directory, then only looks at `~/.eslintrc` if it hasn't found anything yet, though here you probably only want to mention that it goes up to the root directory.

Comment (gyandeeps) [inline docs/user-guide/configuring.md]: Basically it stops at the directory where it finds a `.eslintrc` file with `root` set to `true`.

**Comment by IanVS** [comment]
Anything else I should change before I squash commits?  Is the test sufficient?

File: docs/user-guide/configuring.md [hunk 4]
```diff
@@ -344,7 +344,7 @@ The second way to use configuration files is via `.eslintrc` and `package.json`
 
 In each case, the settings in the configuration file override default settings.
 
-## Configuration Cascading and Hierarchy
+## Configuration Cascading
```
Comment (nzakas) [inline docs/user-guide/configuring.md]: Why the change here?

File: lib/config.js [hunk 5]
```diff
@@ -229,6 +230,16 @@ function getLocalConfig(thisConfig, directory) {
             continue;
         }
 
+        // Check for root flag
+        if (localConfig.root === true) {
+            rootPath = path.dirname(localConfigFile);
```
Comment (nzakas) [inline lib/config.js]: Can't you just `break` here and avoid the ugliness of the next line?

File: docs/user-guide/configuring.md (hunk 4 above)
Comment (IanVS) [inline docs/user-guide/configuring.md]: Because the section was becoming very long, and it seemed to make more sense to break it apart into two sections (added section for Hierarchy here: https://github.com/eslint/eslint/pull/2924/files#diff-f92aa5edd27b97e8ba555fe7b8806310R403).  Can re-combine if you prefer.

File: lib/config.js (hunk 5 above)
Comment (IanVS) [inline lib/config.js]: That was my first thought, but I don't think it works, because there might be a `package.json` with ESLint configuration after an `.eslintrc` file with `root: true`.   Additionally,  I was also not certain I could rely on the order of elements in the `localConfigFiles` array.

**Comment by IanVS** [comment]
Should I be concerned that Appveyor failed?  It looks like it was an Appveyor configuration issue, maybe someone could restart the build on this PR?

**Comment by gyandeeps** [comment]
Please rebase it with master and you should be good to go.

File: docs/user-guide/configuring.md (hunk 4 above)
Comment (nzakas) [inline docs/user-guide/configuring.md]: The problem with splitting them is that they aren't actually two different things. Cascading is a function of the hierarchy.

File: lib/config.js (hunk 5 above)
Comment (nzakas) [inline lib/config.js]: Ah good point. The order is deterministic, but yeah, you wouldn't want to bail on `.eslintrc` if `package.json` has `root` or vice-versa.

File: docs/user-guide/configuring.md (hunk 4 above)
Comment (IanVS) [inline docs/user-guide/configuring.md]: Sorry, my mistake.

**Comment by IanVS** [comment]
OK, maybe finally ready?

File: lib/config.js (hunk 1 above)
Comment (nzakas) [inline lib/config.js]: Still a bit weary of this check. If there some way to be a bit smarter using the `path` module?

Comment (IanVS) [inline lib/config.js]: With my limited experience, I don't know what else to use, unless you'd prefer I add a dependency for `path-is-inside` as @gyandeeps suggests.

Or perhaps this would be safer? (only if the root folder is named `.eslintrc` or `package.json` I think...)

```
if (rootPath && path.dirname(localConfigFile).indexOf(rootPath) === -1) {
```

**Comment by ilyavolodin** [comment]
LGTM

File: lib/config.js (hunk 1 above)
Comment (nzakas) [inline lib/config.js]: I think path-is-inside is a good approach.

Comment (IanVS) [inline lib/config.js]: Okay, I will swap it in tonight.

**Comment by IanVS** [comment]
`path-is-inside` has been added as requested.  Anything else I should look at on this one?

File: lib/config.js [hunk 6]
```diff
@@ -229,6 +231,16 @@ function getLocalConfig(thisConfig, directory) {
...
+        }
+
+        // If root flag is set, don't consider file if it is above root
+        if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
```
Comment (gyandeeps) [inline lib/config.js]: why are you passing in only the dirname? I would just pass the whole path.

Comment (IanVS) [inline lib/config.js]: I'm just trying to follow the API of the `path-is-inside` module.

Usage from `path-is-inside` README:

> First the path being tested; then the potential parent.

I don't know if full filename will work as expected, so I figured it doesn't hurt to normalize to just consider paths (since that's all we care about in this case anyway).

Comment (gyandeeps) [inline lib/config.js]: Sorry got confused there. Thats is correct.

Comment (nzakas) [inline lib/config.js]: Okay, I had an idea that might simplify all of this. What if we said that `root` specified the root config instead of the root directory? That way, we could just `break` out of this loop and everything is nice and simple. Thoughts?

Comment (IanVS) [inline lib/config.js]: That seems tricky tricky to me.  Does `.eslintrc` always come before `package.json`?  If so, then specifying root in `.eslintrc` would prevent configs in `package.json`, but not vice-versa.  That seems confusing and arbitrary to me.

Personally, I like the idea of specifying a root folder, and I don't think the implementation here is overly complex, is it?

Comment (nzakas) [inline lib/config.js]: I always get jittery when extra unnecessary work is being done, there's more potential for error.

I could be over thinking it, though. Nevermind. :)

**Comment by nzakas** [comment]
OK, merging!

**Comment by gyandeeps** [comment]
yay :+1:  :100: 
Thanks @IanVS
